time-based-packet-transmission-offload.md

```diff
@@ -9,27 +9,121 @@ Because of the time-fidelity requirements of this feature, there will not be sof
 
 ## Table of Context
 
+- [Time based Packet Transmission Offload (TPTO)](#time-based-packet-transmission-offload-tpto)
+  - [Table of Context](#table-of-context)
 - [Winsock](#winsock)
+  - [Checking for TPTO Capability](#checking-for-tpto-capability)
+  - [Enable TPTO](#enable-tpto)
+  - [Sending Packets](#sending-packets)
 - [TCPIP](#tcpip)
+  - [Queue](#queue)
+    - [Unknown](#unknown)
 - [NDIS](#ndis)
 - [Appendix](#appendix)
+  - [Linix](#linix)
 
 # Winsock
 
 The proposed Winsock API for TPTO is as follows.
 
+## Checking for TPTO Capability
+
+An app first checks for TPTO support by querying the `SO_TPTO_SUPPORT` socket option.  
+The option value is a `TPTO_SUPPORT_FLAGS` enum with flags describing the supported capabilities:
+
+```C
+typedef enum _TPTO_SUPPORT_FLAGS {
+    TPTO_SUPPORT_FLAG_NONE          = 0x00,
+    TPTO_SUPPORT_FLAG_SUPPORTED     = 0x01,
+} TPTO_SUPPORT_FLAGS;
+```
+
+## Enable TPTO
+setsockopt with SO_TXTIME
+
+## Sending Packets
+The app calls `WSASendMsg`.   
+The app passes ancillary data in the form of `TPTO_ANCILLARY_DATA`
+```C
+typedef struct _TPTO_ANCILLARY_DATA {
+    uint64_t TxTime;
+    uint64_t ClockId;
+    uint8_t  DropIfRate
+} TPTO_ANCILLARY_DATA;
+```
+
+- TxTime to indicate when the associated packet is sent
+- ClockId to indicate which of several system clocks to be used
+- DropIfRate to indicate whether to drop packet if it cannot be transmitted by the given deadline
+
+- sendmsg with control-message header of type SCM_DROP_IF_LATE to simply drop a packet
+- SCM_CLOCKID which clock should be used for packet timing, default is CLOCK_MONOTONIC
+
+
 > **TODO**
 
 # TCPIP
 
 This section describes necessary updates in the Windows network stack to support TPTO.
 
+## Queue
```

Comment:
I actually want to simplify things at the TCP/IP layer a lot. TCP/IP should be dumb, and we make this purely a HW offload feature. If the HW doesn't support it, we'd return NOT_SUPPORTED for the original query API. If the HW support dynamically goes away, we'll just pass the packets down without the metadata; no special queuing or attempt at the SW layer to implement this logic. Windows timers just aren't good enough. Let's not even try.